### ISSUE TITLE
`type` is a `bash`-ism. Exhaust other alternatives before using.

### ISSUE THREAD

**Issue by bracketttc** [megastep/makeself#179]
This use of `type` is fine and should be converted to a shell function.
https://github.com/megastep/makeself/blob/50fb8888e8189f052be447dfc044831a22c137fc/makeself.sh#L643

This use is potentially problematic as it relies on `type` immediately without trying the more standard `which`.
https://github.com/megastep/makeself/blob/50fb8888e8189f052be447dfc044831a22c137fc/makeself.sh#L167

Apparently, `type` is in the [most recent](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_09_01_01) POSIX shell spec; but, historically, it has not been a part of the POSIX shell.

**Comment by realtime-neil** [comment]
In my own environments, I like to do
```
if ! command -v foobar >/dev/null 2>&1; then
    echo "ERROR: missing command: foobar" >&2
    exit 1
fi
```

Does anyone know if/when `command` joined a POSIX standard?

**Comment by realtime-neil** [comment]
Based on some research, I've been able to determine that `type` is indeed a feature of the Bourne Shell, albeit one that was initially absent from the original in [Version 7 Unix](https://en.wikipedia.org/wiki/Version_7_Unix).

Sven Mascheck's "The Traditional Bourne Shell Family" page (https://www.in-ulm.de/~mascheck/bourne/) does a really good job detailing Bourne shell variants and the noteworthy changes thereto. According to Mascheck,

> A summary of the most important changes
> 
> * Version 7 (1979): control structures,  cmd substitution,  `()` and `{}`,  arbitrary variable names,  `trap`,  `eval`,  special parameter substitutions, `case`      
> * System III (1981): `#`,  `[!...]`,  colon parameter substitution,  `set --`
> * SVR1 (1983): `shift n`
> * SVR2 (1984): functions,  built-ins: `unset` `echo` `type`,  redirection for builtins
> * SVR3 (1986): modern `"$@"`, 8-bit clean, `getopts`, functions do not modify positional parameters
> * SVR4.0 (1989): job control
> * SVR4.2 (1992): `read -r`

-- https://www.in-ulm.de/~mascheck/bourne/#variants

Based on this summary, `type` was added to the Bourne shell no later than 1984 with the Unix System V Release 2. Here's a detailed look at what Bourne shell changes made it into SVR2: https://www.in-ulm.de/~mascheck/bourne/#svr2

There's a "Table of contents" that follows the summary and shows the heritage of the various prominent Unix variants:
> * Version 7  (and 2.9BSD) and 3BSD and 4.xBSD (and Sinix 5.20 (ucb universe), SunOS 2) - Ultrix - Dynix 3.2 (ucb) - DomainOS 10.3
> * System III  (and Plexus 1.0) and SINIX 5.20 (sie)
> * SVR1 and Iris 3.7
> * SVR2  (and MV/UX 3) and Ultrix sh5 - Version 8 - SunOS 3 - OSF/1 - HP/UX - Dynix 3.2 (att)
> * SVR3  (and MUNIX V3, SINIX 5.20 (att), SCO Xenix 2.3.4) and SunOS 4 - AIX - IRIX 3/4 - SCO Unix - Interactive Unix - DG/UX 40 - DomainOS 10.3 - Dynix PTX
> * SVR4  (and OSF/1-SysV) and SunOS 5 - Heirloom - Schily - IRIX 5/6 - DG/UX 42 - EP/IX - Reliant UNIX
> * SVR4.2  (UnixWare, OpenUnix) 

Based on this table, we can clearly delineate the Unix variants `makeself` *cannot* support if we allow `makeself` to use `type`.

To put it another way, here's a (probably incomplete) list of post-Version-7-Unix features currently used by `makeself` (ordered earliest to latest):
* System III
  * `#` comment character
  * colon parameter substitution
  * `set --`
* SVR1
  * `shift n`
* SVR2
  * functions
  * `unset`
  * `echo`
  * `type`
  * redirection for builtins

**Comment by realtime-neil** [comment]
@megastep I propose either closing this ticket as "invalid" or substantially changing it. This issue could be adapted into a TODO item to formalize a policy around
* which shell should be supported
* which techniques should be used to look for commands

I would propose, respectively, targeting SRV2 Bourne shell and using the `type` shell builtin to look for commands.

**Comment by megastep** [comment]
I agree and it was the implicit consensus pretty much till now; I'm fine supporting only shells that have `type` available as this goes sufficiently back to cover everything I need it to.
